### PR TITLE
Added FileOutputReporter reporter

### PR DESCRIPTION
**Pull request by sirishaNekkanti**
### Summary
Wdio File Output Reporter is a reporter that logs wdio test output to separate files based on locale, theme, and form-factor
addresses:  https://github.com/cerner/terra-toolkit/issues/634

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,7 +68,7 @@
     "heroku-prebuild": "npx lerna init",
     "jest": "jest",
     "lint:js": "eslint --ext .js,.jsx .",
-    "lint": "npm run lint:js",
+    "lint": "npm run lint:js -- --fix",
```
Comment (yuderekyu) [inline package.json]: I'm not sure if we want to default to automatically fixing lints; my thinking is to leave this be so developers can see what errors occur and then decide what action to take. I'm curious what others think

File: package.json [hunk 2]
```diff
@@ -118,5 +118,8 @@
     "webpack": "^4.43.0",
     "webpack-cli": "^3.3.12",
     "webpack-dev-server": "^3.11.0"
+  },
+  "dependencies": {
+    "strip-ansi": "^6.0.0"
```
Comment (yuderekyu) [inline package.json]: is this dependency necessary for the whole monorepo? we should scope this down to the package that uses this

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+        if (!this.resultJsonObject.output[this.moduleName]) {
+          this.resultJsonObject.output[this.moduleName] = [];
+        }
+        let readableMessage = this.getMessage(runner);
+        const readableArrayMessage = readableMessage.toString().split(',');
+        for (let i = 0; i < readableArrayMessage.length; i + 1) {
+          readableArrayMessage[i] = stripAnsi(`${readableArrayMessage[i]}${endOfLine}`);
+        }
+        readableMessage = readableArrayMessage.join('');
```
Comment (yuderekyu) [inline packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: how about being declarative and using map:
```suggestion
        const messages = readableArrayMessage.map((message) => stripAnsi(`${message}${endOfLine}`).join('');
```

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    this.hasResultsDir();
+  }
+
+  /**
+   * Sets results directory for the test run. Uses the wdio reporterOptions.outputDir if set, otherwise
+   * it outputs to tests?/wdio/reports.
+   * @return null;
+   */
+  setResultsDir() {
+    const { reporterOptions } = this.options;
```
Comment (yuderekyu) [inline packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: can you walk through how `this.options` is being propagated and set? Is it coming from the spec reporter that we're extending? 

I tried to dig into this, and my understanding is when we specify what reporters to use inside configuration, we can pass an option as the second argument, and that options object will propagate into this reporter. Is that totally wrong? 

```js
// wdio config
reporters: [[fileOutputReporter, { dir: 'full-stack' }]],
```

```js
// file output reporter
const { dir } = this.options
```

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    this.printReport(runner);
+  }
+
+  getMessage(runner) {
+    const preface = `[${this.getEnviromentCombo(
+      runner.capabilities,
+      false,
+      runner.isMultiremote,
+    ).trim()} #${runner.cid}]`;
+    const divider = '------------------------------------------------------------------';
```
Comment (yuderekyu) [inline packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: I vote we define the preface and divider variables right before they are used. That way it's easier to keep track of the scope of variables

File: package.json (hunk 1 above)
Comment (sirishaNekkanti) [reply package.json]: I deleted it.

File: package.json (hunk 2 above)
Comment (sirishaNekkanti) [reply package.json]: I moved the dependency to terra-functional-testing

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js (hunk 4 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: yes you are right options come from the configurations, in this case we are not using them so, I removed that section

File: package.json (hunk 2 above)
Comment (sirishaNekkanti) [reply package.json]: https://github.com/cerner/terra-toolkit/pull/662/files#diff-72c2847b50d4152bd910ef25c33dafbd31a04d507329a333036e6226e5e8ef71R77

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js (hunk 4 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: https://github.com/cerner/terra-toolkit/pull/662/files#diff-5aca2141ee738df380d2de0da498a0a17727d935d42de74bbbb00a3046b1febeR41-R47

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js (hunk 5 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: https://github.com/cerner/terra-toolkit/pull/662/files#diff-5aca2141ee738df380d2de0da498a0a17727d935d42de74bbbb00a3046b1febeR145-R147

File: packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js (hunk 3 above)
Comment (sirishaNekkanti) [reply packages/terra-functional-testing/src/reporters/fileOutput-reporter/index.js]: https://github.com/cerner/terra-toolkit/pull/662/files#diff-5aca2141ee738df380d2de0da498a0a17727d935d42de74bbbb00a3046b1febeR168-R169

File: packages/jest-config-terra/src/reporters/verbose-reporter/TerraVerboseReporter.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  log(message) {
+    const readableMessage = `${stripAnsi(message)}${endOfLine}`;
+    this.setTestModule(readableMessage);
+    const { moduleName } = this;
+    if (!this.results.output[moduleName]) {
+      this.results.output[moduleName] = [];
+    }
+    if (!this.unformattedResult[moduleName]) {
+      this.unformattedResult[moduleName] = [];
+    }
```
[the remark this answers is not in the document]
Comment (sirishaNekkanti) [reply packages/jest-config-terra/src/reporters/verbose-reporter/TerraVerboseReporter.js:84]: it seems like the jest reporter files are added to this pr. I will open a new clean pr in few minutes

[the remark this answers is not in the document]
Comment (yuderekyu) [reply packages/jest-config-terra/src/reporters/verbose-reporter/TerraVerboseReporter.js:84]: can you close this one then?

**Pull request closed without merging**